RevenueCatUI/View+PurchaseRestoreCompleted.swift

```diff
@@ -332,6 +385,50 @@ private struct OnPurchaseCancelledModifier: ViewModifier {
 
 }
 
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+struct HandlePurchaseAndRestoreModifier: ViewModifier {
+    let performPurchase: PerformPurchase
+    let performRestore: PerformRestore
+
+    func body(content: Content) -> some View {
+        content
+            .modifier(HandlePurchaseModifier(handler: performPurchase))
+            .modifier(HandleRestoreModifier(handler: performRestore))
+    }
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+private struct HandlePurchaseModifier: ViewModifier {
+
+    let handler: PerformPurchase
+
+    func body(content: Content) -> some View {
+        content
+            .onPreferenceChange(HandlePurchasePreferenceKey.self) { purchaseResultReporter in
+                if let purchaseResultReporter {
+                    self.handler(purchaseResultReporter.storeProduct, purchaseResultReporter)
+                }
+            }
+    }
+
+}
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+private struct HandleRestoreModifier: ViewModifier {
+
+    let handler: PerformRestore
+
+    func body(content: Content) -> some View {
+        content
+            .onPreferenceChange(HandleRestorePreferenceKey.self) { restoreResultReporter in
+                if let restoreResultReporter {
+                    self.handler(restoreResultReporter)
+                }
```

Comment:
Do we need to add a log or something if `restoreResultReporter` is `nil`? Or is that already handled somewhere else? 🤔

Reply:
This method gets triggered when PurchaseHandler.performRestore changes. It starts off as `nil`, and then during `performExternalRestoreLogic` it gets assigned an object and triggers this callback. We don't assign that variable anywhere else - the only we we could get here with a `nil` value is if we later assigned it back to `nil`, which as it stands now would be a programming error on our part (because the only time we change this is when we want to call the external callback, and we shouldn't be doing that if we don't have a callback to send to them...). (In theory we should be able to force-unwrap it here :yolo:.)

I [added a log message](https://github.com/RevenueCat/purchases-ios/pull/3917/commits/ce8b8e05520479867eae8b5e106c3d671a80b81c) to catch any future changes that may be in error.